datalad_neuroimaging/extractors/bids_dataset.py

```diff
@@ -0,0 +1,259 @@
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Metadata extractor for BIDS dataset-level information"""
+import logging
+from uuid import UUID
+from bids import BIDSLayout
+from pathlib import Path
+from datalad_metalad.extractors.base import DataOutputCategory, ExtractorResult, DatasetMetadataExtractor
+from datalad.log import log_progress
+import sys
+from datalad.metadata.definitions import vocabulary_id
+from datalad.utils import assure_unicode
+from typing import Dict, List, Union
+import json
+
+lgr = logging.getLogger('datalad.metadata.extractors.bids_dataset')
+
+
+# Main properties used in BIDS v1.6.0 dataset description
+class BIDSProperties:
+    NAME = "Name"
+    BIDSVERSION = "BIDSVersion"
+    DATASETTYPE = "DatasetType"
+    LICENSE = "License"
+    AUTHORS = "Authors"
+    ACKNOWLEDGEMENTS = "Acknowledgements"
+    HOWTOACKNOWLEDGE = "HowToAcknowledge"
+    FUNDING = "Funding"
+    ETHICSAPPROVALS = "EthicsApprovals"
+    REFSANDLINKS = "ReferencesAndLinks"
+    DATASETDOI = "DatasetDOI"
+    HEDVERSION = "HEDVersion"
+    README = "ReadMe"
+
+# Main entities used in BIDS v1.6.0 filenames
+class BIDSEntities:
+    """"""
+    SUBJECT = 'subject'
+    SESSION = 'session'
+    TASK = 'task'
+    ACQUISITION = 'acquisition'
+    CEAGENT = 'ceagent'
+    RECONSTRUCTION = 'reconstruction'
+    DIRECTION = 'direction'
+    RUN = 'run'
+    PROC = 'proc'
+    MODALITY = 'modality'
+    ECHO = 'echo'
+    FLIP = 'flip'
+    INV = 'inv'
+    MT = 'mt'
+    PART = 'part'
+    RECORDING = 'recording'
+    SPACE = 'space'
+    SUFFIX = 'suffix'
+    SCANS = 'scans'
+    FMAP = 'fmap'
+    DATATYPE = 'datatype'
+    EXTENSION = 'extension'
+
+BIDSCONTEXT = {
+    "@id": "https://doi.org/10.5281/zenodo.3686061",
+    'description': 'ad-hoc vocabulary for the Brain Imaging Data Structure (BIDS) standard v1.6.0',
+    'type': vocabulary_id,
+}
+
+DATASET = 'dataset'
+
+BIDS_PROPERTIES_MAPPING= {
+    BIDSProperties.NAME: BIDSProperties.NAME,
+    BIDSProperties.BIDSVERSION: BIDSProperties.BIDSVERSION,
+    BIDSProperties.DATASETTYPE: BIDSProperties.DATASETTYPE,
+    BIDSProperties.LICENSE: BIDSProperties.LICENSE,
+    BIDSProperties.AUTHORS: BIDSProperties.AUTHORS,
+    BIDSProperties.ACKNOWLEDGEMENTS: BIDSProperties.ACKNOWLEDGEMENTS,
+    BIDSProperties.HOWTOACKNOWLEDGE: BIDSProperties.HOWTOACKNOWLEDGE,
+    BIDSProperties.FUNDING: BIDSProperties.FUNDING,
+    BIDSProperties.ETHICSAPPROVALS: BIDSProperties.ETHICSAPPROVALS,
+    BIDSProperties.REFSANDLINKS: BIDSProperties.REFSANDLINKS,
+    BIDSProperties.DATASETDOI: BIDSProperties.DATASETDOI,
+    BIDSProperties.HEDVERSION: BIDSProperties.HEDVERSION,
+}
```

Comment:
probably too early to whine but my duplication allergy kicks in:  what are those and why they are needed and why they aren't just smth like

```python
BIDS_PROPERTIES_MAPPING = { getattr(BIDSProperties, a): getattr(BIDSProperties, a) for a in dir(BIDSProperties) }
```
if really needed?

Reply:
No this is a good whine :)

So the mapping dict originally had values taken from a different class, resulting in BIDS properties being mapped to fields with different names/structures. But during development I changed this and decided (for the time being) on a one-to-one mapping between the extracted metadata and the metadata being represented in the output. However, I left the mapping dict as is in case, during reviewing, we do in fact decide that a different mapping makes more sense.

Therefore, if we stick to the current mapping procedure, then I agree we can cure your duplication allergy :)

Reply:
Keeping these duplications for now since I predict that mappings will still be useful when new BIDS versions update the BIDS entities/variables and assuming the extractor doesn't necessarily want to follow BIDS as strictly.